### ISSUE TITLE
Differentiability wrt. to mesh vertex positions

### ISSUE THREAD

**Issue by andrinr** [tum-pbs/PhiFlow#187]
I'm currently trying to figure out whether phiflow supports differentiability wrt to mesh vertex positions. My initial assessment of the way the mesh is setup and initial failed experiments make me believe this is not supported.

**Comment by holl-** [comment]
When using the `Mesh` class, the vertex positions are variable by default. So a function like `loss_function(mesh: Mesh)` can be differentiated w.r.t. the vertex positions. When you pass it as part of a `Field`, make sure to set `variable_attrs=('values', 'geometry')`, else the vertex positions are taken as constant.

**Comment by dionhaefner** [comment]
I'm trying to adapt [the cylinder FVM example](https://github.com/tum-pbs/PhiFlow/blob/develop/examples/mesh/FVM_Cylinder_GMsh.ipynb) to include a simple derivative wrt. mesh vertices, but running into several issues.

- https://github.com/tum-pbs/PhiFlow/issues/184 triggers for both install via `pip` and from `master`. Going with development branches for now.
- Running the example with JAX instead of torch (`from phi.flow import *`) diverges immediately with this exception:

```python
File ~/.virtualenvs/science/lib/python3.12/site-packages/phiml/math/_optimize.py:199, in SolveInfo.convergence_check(self, only_warn)
    197             warnings.warn(self.msg, ConvergenceWarning)
    198         else:
--> 199             raise Diverged(self)
    200 if not self.converged.trajectory[-1].all:
    201     if NotConverged not in self.solve.suppress:

Diverged: Solve diverged within -1 iterations using scipy.sparse.linalg.spsolve.
```

Any suggestions on how to fix it, or another example that we could use for prototyping that uses an unstructured mesh?

**Comment by holl-** [comment]
Hey @dionhaefner Could you update to the develop branch?
```
pip uninstall phiflow
pip install git+https://github.com/tum-pbs/PhiFlow@develop
pip uninstall phiml
pip install git+https://github.com/tum-pbs/PhiML@develop
```
That should hopefully fix the errors.

**Comment by holl-** [comment]
To verify that the gradient w.r.t. vertex positions works in principle, you can start with this:

```python
mesh = geom.load_gmsh('cylinder.msh', ('y-', 'x+', 'y+', 'x-', 'cyl+', 'cyl-'))

def loss_function(mesh: Mesh):
    return math.l2_loss(mesh.center)

d_mesh: Mesh = gradient(loss_function, 'mesh', False)(mesh)
show(PointCloud(mesh.vertices, -.05 * d_mesh.vertices.center))
```

Yeah, the divergence error looks like a numerics problem.
To get it running now, you can employ the following hack
```python
v, p = fluid.make_incompressible(v, (), Solve('scipy-direct', suppress=[Diverged, NotConverged]))
```
Not sure why the solve doesn't converge, it should be easily solveable with a ZERO_GRADIENT boundary condition...

**Comment by holl-** [comment]
Here is an example that derives vertex gradients based on a fluid simulation (with some caveats).

<details> <summary>Demo Script</summary>

```python
mesh = geom.load_gmsh('cylinder.msh', ('y-', 'x+', 'y+', 'x-', 'cyl+', 'cyl-'))
boundary = {'x-': vec(x=1, y=0), 'x+': ZERO_GRADIENT, 'y': 0, 'cyl': 0}

@jit_compile_linear
def momentum_eq(u, u_prev, dt, diffusivity=0.01):
    diffusion_term = dt * diffuse.differential(u, diffusivity, correct_skew=False)
    advection_term = dt * advect.differential(u, u_prev, order=1)
    return u + advection_term + diffusion_term

# @jit_compile
def implicit_time_step(v, dt, p=0):
    v, p = fluid.make_incompressible(v, (), Solve('scipy-direct', abs_tol=0, suppress=[NotConverged, Diverged]))
    v = math.solve_linear(momentum_eq, v, Solve('scipy-lsqr', x0=v, abs_tol=0), u_prev=v, dt=-dt, grad_for_f=True)
    return v, p


def loss_fn(mesh: Mesh):
    v = Field(mesh, tensor(vec(x=0, y=0)), boundary)
    for _ in trange(2):
        v, p = implicit_time_step(v, dt=0.001)
    loss = field.mean(-v['x'] * mask)
    return loss, v, p


mask = Field(mesh, lambda x,y: x > 2, boundary)  # loss is computed in this region

grad_fn = gradient(loss_fn, 'mesh')
(loss, v, p), d_mesh = grad_fn(mesh)
print(d_mesh.vertex_positions)
show(PointCloud(mesh.vertex_positions, -.05 * d_mesh.vertex_positions))
```

</details>

Note the `grad_for_f=True` which enables backpropagation through the matrix entries of the linear system.

**Caveat 1**: This only runs on `develop` at the moment
```
pip uninstall phiflow
pip install git+https://github.com/tum-pbs/PhiFlow@develop
pip uninstall phiml
pip install git+https://github.com/tum-pbs/PhiML@develop
``` 

**Caveat 2**: The gradients become unstable for more time steps. I think this is related to the outer boundary which produces the largest gradients. The solution may be to set gradient at the outer boundary to zero manually for every time step during backprop. You can do this by defining an identity forward function with a custom gradient using [`custom_gradient()`](https://tum-pbs.github.io/PhiML/phiml/math/index.html#phiml.math.custom_gradient)

**Caveat 3**: JIT-compilation with JAX is broken at the moment for the gradients. This can be fixed, but these kinds of errors are always hard to debug. I'll look into it if you think Φ-Flow suits your needs otherwise.

**Comment by andrinr** [comment]
@holl- lightspeed! Will check it out right away 👍
